src/Microsoft.Azure.WebJobs.Host/Loggers/Logger/ApplicationInsights/ApplicationInsightsLoggerProvider.cs

```diff
@@ -13,22 +13,47 @@ internal class ApplicationInsightsLoggerProvider : ILoggerProvider
         private readonly TelemetryClient _client;
         private readonly Func<string, LogLevel, bool> _filter;
 
-        public ApplicationInsightsLoggerProvider(string instrumentationKey, Func<string, LogLevel, bool> filter,
-            ITelemetryClientFactory clientFactory, SamplingPercentageEstimatorSettings samplingSettings)
+        private ITelemetryClientFactory _clientFactory;
+        private bool _disposed;
+
+        public ApplicationInsightsLoggerProvider(Func<string, LogLevel, bool> filter, ITelemetryClientFactory clientFactory)
         {
-            if (instrumentationKey == null)
+            if (clientFactory == null)
             {
-                throw new ArgumentNullException(nameof(instrumentationKey));
+                throw new ArgumentNullException(nameof(clientFactory));
             }
 
-            _client = clientFactory.Create(instrumentationKey, samplingSettings);
+            _clientFactory = clientFactory;
+            _client = _clientFactory.Create();
             _filter = filter;
         }
 
         public ILogger CreateLogger(string categoryName) => new ApplicationInsightsLogger(_client, categoryName, _filter);
 
         public void Dispose()
         {
+            if (!_disposed)
+            {
+                _clientFactory.Dispose();
+
+                if (_client != null)
+                {
+                    try
+                    {
+                        _client.Flush();
+
+                        // This sleep isn't ideal, but the flush is async so it's the best we have right now. This is
+                        // being tracked at https://github.com/Microsoft/ApplicationInsights-dotnet/issues/407
+                        Thread.Sleep(2000);
```

Comment:
Do we definitely have data loss without the sleep? Have you tried different values?